irutil: add instruction and terminator operand use-tracking and value update API

mewmew opened the pull request:
Rather than adding the use-tracking to llir/llvm/ir, we currently add
this to irutil so we may get more experience with the API and see if its
the right fit for instruction and terminator operand use-tracking and
value update.

Note: this is the stronger version of the API, as outlined by @pwaller
in https://github.com/llir/llvm/issues/42#issue-383892282

Notably, operands are tracked by pointer (*value.Value) which also enables
updating values using the same API.

Fixes llir/llvm#42.

Updates llir/llvm#19 (note, we don't do deep use tracking in this PR, so llir/llvm#19 is only partially fixed).

cc: @pwaller and @dannypsnl

mewmew committed:
irutil: add instruction and terminator operand use-tracking and value update API

Rather than adding the use-tracking to llir/llvm/ir, we currently add
this to irutil so we may get more experience with the API and see if its
the right fit for instruction and terminator operand use-tracking and
value update.

Note: this is the stronger version of the API, as outlined by @pwaller
in https://github.com/llir/llvm/issues/42#issue-383892282

Notably, operands are tracked by pointer (*value.Value) which also enables
updating values using the same API.

Fixes llir/llvm#42.

mewmew commented on go.mod:
```diff
@@ -3,3 +3,5 @@ module github.com/llir/irutil
 
 require github.com/llir/llvm v0.3.0-pre7.0.20191220191840-fc6d0e8d9bc6
+
+replace github.com/llir/llvm => github.com/mewpull/llvm v0.3.0-pre7.0.20191221041541-3f58af5a0014
```
Note, this PR depends on https://github.com/llir/llvm/pull/122. After llir/llvm#122 has been merged (if consensus is reached that it should be merged), then we will update `go.mod` to remove this replacement directive and use `llir/llvm` directly.

mewmew replied:
> Note, this PR depends on llir/llvm#122. After llir/llvm#122 has been merged (if consensus is reached that it should be merged), then we will update go.mod to remove this replacement directive and use llir/llvm directly.

Done. Updated to use `llir/llvm`.

mewmew committed:
irutil: add preliminary support for Comment pseudo-instruciton in InstUses

Note: we should try to arrive at a more general way of handling custom
user-defined instructions.

pwaller commented on use.go:
```diff
@@ -0,0 +1,464 @@
+// TODO: consider renaming InstUses to InstOperands.
+
+// InstUses returns value uses of the given instruction.
+func InstUses(inst ir.Instruction) []*Use {
```
I have limited time right now, and may not be able to get to all the things you want me to look at.

API-wise, I wanted to note that this will result in allocation for all calls. I can imagine use cases where this might be called a lot.

The way to avoid it would be to have:

```go
func InstUses(backing []*Use, inst ir.Instruction) (out []*Use)
```

Then, append to the backing.

Then, you'll get allocation amortized over the caller. If the caller can't be bothered to set up a buffer to append to, they can just pass `nil` as the backing pointer, and `append` will allocate, as before.

pwaller replied:
Note: `InstUses` reads to me like it will find all uses of an instruction, rather than things which the instruction makes use of. Don't have something immediately in mind which fixes that problem.

mewmew replied:
> API-wise, I wanted to note that this will result in allocation for all calls. I can imagine use cases where this might be called a lot.

Fair point. I was quite certain this would be one of your feedbacks :) I try to strive for the most user-friendly API first, but I agree with you. Given that these functions may be called a lot, it makes sense to make it possible to avoid unnecessary allocations.

> Then, you'll get allocation amortized over the caller. If the caller can't be bothered to set up a buffer to append to, they can just pass nil as the backing pointer, and append will allocate, as before.

Sounds good. I'll update the API as suggested.

> Note: InstUses reads to me like it will find all uses of an instruction, rather than things which the instruction makes use of. Don't have something immediately in mind which fixes that problem.

I was considering to use the name `InstOperands`. The downside is that it doesn't mirror well to `FuncOperands`.

Another idea would then be to use `UsesInInst`, `UsesInFunc`. Can't say I prefer this over `InstUses`, as I think it's one of those things you get used to. However, if there is more people who prefer an alternative name to `InstUses` (and `FuncUses`), then I'd be more than happy to change it.

Also, feel free to suggest other names that I have not yet listed above. I am easily persuaded if we just have a good proposal for what to call it instead.

mewmew replied:
> I have limited time right now, and may not be able to get to all the things you want me to look at.

No worries @pwaller, I was guessing you'd be busy during the winter holidays ^^

Just a quick question to move things forward. Are you ok with the change proposed in llir/llvm#122? If so, I'll merge that PR first, as it is a dependency of this PR.

mewmew replied:
Thanks for the suggested change to the API @pwaller. It has now been applied.

The API has now been updated to make use of a backing slice to avoid memory allocation. To further reduce memory allocations, the use-tracking API has been updated to operate on `[]Use` instead of `[]*Use`. We still track a pointer to the used value, so this update has no effect on the usability of the API, it only helps to reduce memory allocations.

mewmew committed:
irutil: add backing slice to avoid memory allocation in use-tracking API

As suggested by @pwaller in https://github.com/llir/irutil/pull/1#discussion_r360733838

mewmew committed:
irutil: return backing slice also from instructions which have no value operands

Without this change, repeated invokations of InstUses (as used in FuncUses)
would result in incorrect results.

	for _, inst := range block.Insts {
		backing = InstUses(backing, inst)
	}

mewmew committed:
irutil: refine use-tracking API to operate on []Use instead of []*Use

Since we already track a pointer to the used value in the Use structure,
we can pass Use by value. This should help reduce memory allocations.

pwaller commented:
Further thoughts.

I think that `InstUses` as you have it here is really `interface { OperandsAsUses([]Use) []Use }`, and should be methods on the individual types. 

0. Including the word `Operands` somewhere in the name eliminates the confusion with `Uses`. In the LLVM API "`Inst.Uses()`" means to obtain where the instruction is used, not to make a list of uses out of the operands of the instruction as I understand we're doing here (please correct me if there is an edge case I've missed).
1. Having a method on the type makes it clearer that it is a property of instances of the type, as opposed to some algorithm which will compute the uses of an instruction.
2. This allows dynamic dispatch, rather than at runtime having to iterate through all the possible types, you can directly call just the relevant code (as I understand it, but I've not looked and done the measurements, so I could be mistaken - caveat emptor).
3. This solves the "custom" case. If anyone wants to implement `OperandsAsUses`, they can.

I realise this means that `type Use` would have to go in the IR package, and you're reluctant to put it there. For what it's worth, `Value` and `Use` live in the same library in llvm, and this makes some sense to me. The definition of `Use` is so small and so often useful, I don't currently see why you'd want to put it in a different package.

I envisage a `use.go` in the `ir` package which contains all of the use methods for the different types.

A note on naming: `backing` came to me very rapidly. I think I might prefer `in` and `out` in this usecase:

```go
func foo(in []bar) (out []bar) {
  out = in
  ...
  out = append(out, extra...)
}
```

... but please adjust to your (good) taste! :)

The other thing in the back of my mind is - how are we going to maintain the sets of `UsesOf Inst => []Use`? Especially once you start manipulating the IR. In LLVM I understand that each value maintains a `UseList` within it.

mewmew committed:
irutil: use llir/llvm (instead of mewpull/llvm) now that llir/llvm#122 is merged

mewmew committed:
irutil: add dead code elimination as use-tracking API example

Note: this implementation of dce is a toy example. It does not take deep
value uses into account, and as such, cannot handle use tracking when
local variables are used within constant expressions.

This issue would be resolved when irutil.FuncUses is extended to handle
deep value use tracking.

mewmew commented:
I've added an example use case of the API to get a better feel for the use-tracking API.

For use-tracking, see the dead code elimination example at: https://github.com/llir/irutil/blob/503ea997aa34863b1a261bd5c0a93a46245cb371/example_test.go

And for used value modification, see the NOP `trunc` example at: https://gist.github.com/mewmew/05d625cc101d81e4f7b79469f41105fe

mewmew committed:
irutil: check error of dce in example

mewmew committed:
irutil: minor update of spelling

mewmew commented:
> I think that InstUses as you have it here is really interface { OperandsAsUses([]Use) []Use }, and should be methods on the individual types.

Indeed, `{Inst,Term}Uses` is really `interface { OperandsAsUses([]Use) []Use }`.

> In the LLVM API "Inst.Uses()" means to obtain where the instruction is used, not to make a list of uses out of the operands of the instruction as I understand we're doing here

Yeah, using `InstUses` creates a confustion as it's too close to `Inst.Uses()` of the official LLVM API. Your proposed `OperandsAsUses` makes sense and I like it. And as you predicted, I don't want to put it in `llir/llvm/ir` *just yet*.

> Having a method on the type makes it clearer that it is a property of instances of the type, as opposed to some algorithm which will compute the uses of an instruction.

Agreed. Using methods for `OperandsAsUses` is most likely what we would do for the v0.4 release proper. Until then, I think we can live with `func irutil.InstOperandsAsUses(in []Use, inst ir.Instruction)` (and analogous for `TermOperandsAsUses`). It is a bit ugly, but lets us experiment and get a feel for the API. I know this doesn't solve the "custom" instructions issue, but I imagine we can introduce those functions in irutil, and still add the interface `interface { OperandsAsUses([]Use) []Use }`, which we could then use to check "custom" instructions.

> This allows dynamic dispatch, rather than at runtime having to iterate through all the possible types, you can directly call just the relevant code (as I understand it, but I've not looked and done the measurements, so I could be mistaken - caveat emptor).

Yeah, methods is most likely the way to go, for v0.4 that is :)

> This solves the "custom" case. If anyone wants to implement OperandsAsUses, they can.

Thanks for sharing your thoughts on this Peter! I think it definitely helped us move closer to a good use-tracking API, and solves the already identified issue of "custom" instructions ("custom" anything use-tracking related really, e.g. perhaps a "custom" constant expression).

> I envisage a use.go in the ir package which contains all of the use methods for the different types.

Agreed.

> The other thing in the back of my mind is - how are we going to maintain the sets of UsesOf Inst => []Use? Especially once you start manipulating the IR. In LLVM I understand that each value maintains a UseList within it.

I have no idea yet :) That's essentially the main reason I want to keep the use-tracking API in `irutil` for now, as it's experimental, and only by iterative design we will arrive at something solid enough to be "promoted" into `llir/llvm` proper. That would coincide with the v0.4 release of `llir/llvm`.

mewmew committed:
irutil: update to latest version of llir/llvm

mewmew committed:
irutil: update to v0.3.0 of llir/llvm
